File: data/verified_projects/commons-pool/src/test/org/apache/commons/pool2/TestKeyedObjectPool.py
import pytest

import unittest
from abc import ABC, abstractmethod
from src.test.org.apache.commons.pool2.MethodCall import *
from src.test.org.apache.commons.pool2.PrivateException import *
from src.main.org.apache.commons.pool2.KeyedObjectPool import *
from typing import List, Any

class TestKeyedObjectPool(unittest.TestCase, ABC):

    class FailingKeyedPooledObjectFactory:

        def __init__(self) -> None:
            self.__methodCalls = []
            self.__count = 0
            self.__makeObjectFail = False
            self.__activateObjectFail = False
            self.__validateObjectFail = False
            self.__passivateObjectFail = False
            self.__destroyObjectFail = False

        
        def activateObject(self, key, obj) -> None:
            try:
                self.__methodCalls.append(
                    MethodCall.MethodCall0("activateObject", key, obj.getObject())
                )
                if self.__activateObjectFail:
                    raise PrivateException("activateObject")
            except Exception as e:
                raise e

        
        def destroyObject(self, key, obj) -> None:
            try:
                self.__methodCalls.append(
                    MethodCall.MethodCall0("destroyObject", key, obj.getObject())
                )
                if self.__destroyObjectFail:
                    raise PrivateException("destroyObject")
            except Exception as e:
                raise e

        
        def getCurrentCount(self) -> int:
            return self.__count

        
        def getMethodCalls(self) -> List[MethodCall]:
            return self.__methodCalls

        
        def isActivateObjectFail(self) -> bool:
            return self.__activateObjectFail

        
        def isDestroyObjectFail(self) -> bool:
            return self.__destroyObjectFail

        
        def isMakeObjectFail(self) -> bool:
            return self.__makeObjectFail

        
        def isPassivateObjectFail(self) -> bool:
            return self.__passivateObjectFail

        
        def isValidateObjectFail(self) -> bool:
            return self.__validateObjectFail

        
        def passivateObject(self, key, obj) -> None:
            self.__methodCalls.append(
                MethodCall.MethodCall0("passivateObject", key, obj.getObject())
            )
            if self.__passivateObjectFail:
                raise PrivateException("passivateObject")

        
        def reset(self) -> None:
            self.__count = 0
            self.__methodCalls.clear()
            self.setMakeObjectFail(False)
            self.setActivateObjectFail(False)
            self.setValidateObjectFail(False)
            self.setPassivateObjectFail(False)
            self.setDestroyObjectFail(False)

        
        def setActivateObjectFail(self, activateObjectFail) -> None:
            self.__activateObjectFail = activateObjectFail

        
        def setCurrentCount(self, count) -> None:
            self.__count = count

        
        def setDestroyObjectFail(self, destroyObjectFail) -> None:
            self.__destroyObjectFail = destroyObjectFail

        
        def setMakeObjectFail(self, makeObjectFail) -> None:
            self.__makeObjectFail = makeObjectFail

        
        def setPassivateObjectFail(self, passivateObjectFail) -> None:
            self.__passivateObjectFail = passivateObjectFail

        
        def setValidateObjectFail(self, validateObjectFail) -> None:
            self.__validateObjectFail = validateObjectFail

        
        def validateObject(self, key, obj) -> bool:
            call = MethodCall.MethodCall0("validateObject", key, obj.getObject())
            self.__methodCalls.append(call)
            if self.__validateObjectFail:
                raise PrivateException("validateObject")
            r = True
            call.returned(bool(r))
            return r

    
    _KEY = "key"

    
    __test__ = False

    
    __pool = None

    __ZERO = int(0)

    __ONE = int(1)

    
    def __clear(self, factory, expectedMethods) -> None:
        factory.getMethodCalls().clear()
        expectedMethods.clear()
    

    @abstractmethod
    def _getNthObject(self, key, n) -> Any:
        pass

    
    @abstractmethod
    def _isFifo(self) -> bool:
        pass

    
    @abstractmethod
    def _isLifo(self) -> bool:
        pass

    
    @abstractmethod
    def _makeEmptyPool0(self, minCapacity) -> KeyedObjectPool:
        pass

    
    @abstractmethod
    def _makeEmptyPool1(self, factory) -> KeyedObjectPool:
        pass

    
    @abstractmethod
    def _makeKey(self, n) -> Any:
        pass

    
    def __reset(self, pool, factory, expectedMethods) -> None:
        try:
            pool.clear0()
            self.__clear(factory, expectedMethods)
            factory.reset()
        except Exception as e:
                raise e
    
    
    def tearDown(self) -> None:
        self.__pool = None

    @pytest.mark.test
    def testBaseAddObject(self) -> None:
        try:
            try:
                self.__pool = self._makeEmptyPool0(3)
            except (RuntimeError, NotImplementedError):
                return

            key = self._makeKey(0)
            try:
                self.assertEqual(0, self.__pool.getNumIdle0())
                self.assertEqual(0, self.__pool.getNumActive0())
                self.assertEqual(0, self.__pool.getNumIdle1(key))
                self.assertEqual(0, self.__pool.getNumActive1(key))
                self.__pool.addObject(key)
                self.assertEqual(1, self.__pool.getNumIdle0())
                self.assertEqual(0, self.__pool.getNumActive0())
                self.assertEqual(1, self.__pool.getNumIdle1(key))
                self.assertEqual(0, self.__pool.getNumActive1(key))
                obj = self.__pool.borrowObject(key)
                self.assertEqual(self._getNthObject(key, 0), obj)
                self.assertEqual(0, self.__pool.getNumIdle0())
                self.assertEqual(1, self.__pool.getNumActive0())
                self.assertEqual(0, self.__pool.getNumIdle1(key))
                self.assertEqual(1, self.__pool.getNumActive1(key))
                self.__pool.returnObject(key, obj)
                self.assertEqual(1, self.__pool.getNumIdle0())
                self.assertEqual(0, self.__pool.getNumActive0())
                self.assertEqual(1, self.__pool.getNumIdle1(key))
                self.assertEqual(0, self.__pool.getNumActive1(key))
            except (RuntimeError, NotImplementedError):
                return
            finally:
                self.__pool.close()
        except Exception as e:
            self.fail(f"An unexpected exception occurred: {e}")

    
    @pytest.mark.test
    def testBaseBorrow(self) -> None:
        try:
            try:
                self.__pool = self._makeEmptyPool0(3)
            except (RuntimeError, NotImplementedError):
                return
            
            keya = self._makeKey(0)
            keyb = self._makeKey(1)
            self.assertEqual(
                self._getNthObject(keya, 0),
                self.__pool.borrowObject(keya),
                "1"
            )
            self.assertEqual(
                self._getNthObject(keyb, 0),
                self.__pool.borrowObject(keyb),
                "2"
            )
            self.assertEqual(
                self._getNthObject(keyb, 1),
                self.__pool.borrowObject(keyb),
                "3"
            )
            self.assertEqual(
                self._getNthObject(keya, 1),
                self.__pool.borrowObject(keya),
                "4"
            )
            self.assertEqual(
                self._getNthObject(keyb, 2),
                self.__pool.borrowObject(keyb),
                "5"
            )
            self.assertEqual(
                self._getNthObject(keya, 2),
                self.__pool.borrowObject(keya),
                "6"
            )
            self.__pool.close()
        except Exception as e:
            self.fail(f"An unexpected exception occurred: {e}")

    
    @pytest.mark.test
    def testBaseBorrowReturn(self) -> None:
        try:
            try:
                self.__pool = self._makeEmptyPool0(3)
            except (RuntimeError, NotImplementedError):
                return

            keya = self._makeKey(0)
            obj0 = self.__pool.borrowObject(keya)
            self.assertEqual(self._getNthObject(keya, 0), obj0)
            obj1 = self.__pool.borrowObject(keya)
            self.assertEqual(self._getNthObject(keya, 1), obj1)
            obj2 = self.__pool.borrowObject(keya)
            self.assertEqual(self._getNthObject(keya, 2), obj2)
            self.__pool.returnObject(keya, obj2)
            obj2 = self.__pool.borrowObject(keya)
            self.assertEqual(self._getNthObject(keya, 2), obj2)
            self.__pool.returnObject(keya, obj1)
            obj1 = self.__pool.borrowObject(keya)
            self.assertEqual(self._getNthObject(keya, 1), obj1)
            self.__pool.returnObject(keya, obj0)
            self.__pool.returnObject(keya, obj2)
            obj2 = self.__pool.borrowObject(keya)
            if self._isLifo():
                self.assertEqual(self._getNthObject(keya, 2), obj2)
            if self._isFifo():
                self.assertEqual(self._getNthObject(keya, 0), obj2)
            obj0 = self.__pool.borrowObject(keya)
            if self._isLifo():
                self.assertEqual(self._getNthObject(keya, 0), obj0)
            if self._isFifo():
                self.assertEqual(self._getNthObject(keya, 2), obj0)
            self.__pool.close()
        except Exception as e:
            self.fail(f"An unexpected exception occurred: {e}")

    
    @pytest.mark.test
    def testBaseClear(self) -> None:
        try:
            try:
                self.__pool = self._makeEmptyPool0(3)
            except (RuntimeError, NotImplementedError):
                return

            keya = self._makeKey(0)
            self.assertEqual(0, self.__pool.getNumActive1(keya))
            self.assertEqual(0, self.__pool.getNumIdle1(keya))
            obj0 = self.__pool.borrowObject(keya)
            obj1 = self.__pool.borrowObject(keya)
            self.assertEqual(2, self.__pool.getNumActive1(keya))
            self.assertEqual(0, self.__pool.getNumIdle1(keya))
            self.__pool.returnObject(keya, obj1)
            self.__pool.returnObject(keya, obj0)
            self.assertEqual(0, self.__pool.getNumActive1(keya))
            self.assertEqual(2, self.__pool.getNumIdle1(keya))
            self.__pool.clear1(keya)
            self.assertEqual(0, self.__pool.getNumActive1(keya))
            self.assertEqual(0, self.__pool.getNumIdle1(keya))
            obj2 = self.__pool.borrowObject(keya)
            self.assertEqual(self._getNthObject(keya, 2), obj2)
            self.__pool.close()
        except Exception as e:
            self.fail(f"An unexpected exception occurred: {e}")

    
    @pytest.mark.test
    def testBaseInvalidateObject(self) -> None:
        try:
            try:
                self.__pool = self._makeEmptyPool0(3)
            except (RuntimeError, NotImplementedError):
                return

            keya = self._makeKey(0)
            self.assertEqual(0, self.__pool.getNumActive1(keya))
            self.assertEqual(0, self.__pool.getNumIdle1(keya))
            obj0 = self.__pool.borrowObject(keya)
            obj1 = self.__pool.borrowObject(keya)
            self.assertEqual(2, self.__pool.getNumActive1(keya))
            self.assertEqual(0, self.__pool.getNumIdle1(keya))
            self.__pool.invalidateObject0(keya, obj0)
            self.assertEqual(1, self.__pool.getNumActive1(keya))
            self.assertEqual(0, self.__pool.getNumIdle1(keya))
            self.__pool.invalidateObject0(keya, obj1)
            self.assertEqual(0, self.__pool.getNumActive1(keya))
            self.assertEqual(0, self.__pool.getNumIdle1(keya))
            self.__pool.close()
        except Exception as e:
            self.fail(f"An unexpected exception occurred: {e}")

    
    @pytest.mark.test
    def testBaseNumActiveNumIdle(self) -> None:
        try:
            try:
                self.__pool = self._makeEmptyPool0(3)
            except (RuntimeError, NotImplementedError):
                return

            keya = self._makeKey(0)
            self.assertEqual(0, self.__pool.getNumActive1(keya))
            self.assertEqual(0, self.__pool.getNumIdle1(keya))
            obj0 = self.__pool.borrowObject(keya)
            self.assertEqual(1, self.__pool.getNumActive1(keya))
            self.assertEqual(0, self.__pool.getNumIdle1(keya))
            obj1 = self.__pool.borrowObject(keya)
            self.assertEqual(2, self.__pool.getNumActive1(keya))
            self.assertEqual(0, self.__pool.getNumIdle1(keya))
            self.__pool.returnObject(keya, obj1)
            self.assertEqual(1, self.__pool.getNumActive1(keya))
            self.assertEqual(1, self.__pool.getNumIdle1(keya))
            self.__pool.returnObject(keya, obj0)
            self.assertEqual(0, self.__pool.getNumActive1(keya))
            self.assertEqual(2, self.__pool.getNumIdle1(keya))

            self.assertEqual(0, self.__pool.getNumActive1("xyzzy12345"))
            self.assertEqual(0, self.__pool.getNumIdle1("xyzzy12345"))

            self.__pool.close()
        except Exception as e:
            self.fail(f"An unexpected exception occurred: {e}")

    
    @pytest.mark.test
    def testBaseNumActiveNumIdle2(self) -> None:
        try:
            try:
                self.__pool = self._makeEmptyPool0(6)
            except (RuntimeError, NotImplementedError):
                return
            
            keya = self._makeKey(0)
            keyb = self._makeKey(1)
            self.assertEqual(0, self.__pool.getNumActive0())
            self.assertEqual(0, self.__pool.getNumIdle0())
            self.assertEqual(0, self.__pool.getNumActive1(keya))
            self.assertEqual(0, self.__pool.getNumIdle1(keya))
            self.assertEqual(0, self.__pool.getNumActive1(keyb))
            self.assertEqual(0, self.__pool.getNumIdle1(keyb))

            objA0 = self.__pool.borrowObject(keya)
            objB0 = self.__pool.borrowObject(keyb)

            self.assertEqual(2, self.__pool.getNumActive0())
            self.assertEqual(0, self.__pool.getNumIdle0())
            self.assertEqual(1, self.__pool.getNumActive1(keya))
            self.assertEqual(0, self.__pool.getNumIdle1(keya))
            self.assertEqual(1, self.__pool.getNumActive1(keyb))
            self.assertEqual(0, self.__pool.getNumIdle1(keyb))

            objA1 = self.__pool.borrowObject(keya)
            objB1 = self.__pool.borrowObject(keyb)

            self.assertEqual(4, self.__pool.getNumActive0())
            self.assertEqual(0, self.__pool.getNumIdle0())
            self.assertEqual(2, self.__pool.getNumActive1(keya))
            self.assertEqual(0, self.__pool.getNumIdle1(keya))
            self.assertEqual(2, self.__pool.getNumActive1(keyb))
            self.assertEqual(0, self.__pool.getNumIdle1(keyb))

            self.__pool.returnObject(keya, objA0)
            self.__pool.returnObject(keyb, objB0)

            self.assertEqual(2, self.__pool.getNumActive0())
            self.assertEqual(2, self.__pool.getNumIdle0())
            self.assertEqual(1, self.__pool.getNumActive1(keya))
            self.assertEqual(1, self.__pool.getNumIdle1(keya))
            self.assertEqual(1, self.__pool.getNumActive1(keyb))
            self.assertEqual(1, self.__pool.getNumIdle1(keyb))

            self.__pool.returnObject(keya, objA1)
            self.__pool.returnObject(keyb, objB1)

            self.assertEqual(0, self.__pool.getNumActive0())
            self.assertEqual(4, self.__pool.getNumIdle0())
            self.assertEqual(0, self.__pool.getNumActive1(keya))
            self.assertEqual(2, self.__pool.getNumIdle1(keya))
            self.assertEqual(0, self.__pool.getNumActive1(keyb))
            self.assertEqual(2, self.__pool.getNumIdle1(keyb))

            self.__pool.close()

        except Exception as e:
            self.fail(f"An unexpected exception occurred: {e}")

File: java_projects/cleaned_final_projects_evosuite/commons-pool/src/test/java/org/apache/commons/pool2/TestKeyedObjectPool.java

/*
  Licensed to the Apache Software Foundation (ASF) under one or more
  contributor license agreements.  See the NOTICE file distributed with
  this work for additional information regarding copyright ownership.
  The ASF licenses this file to You under the Apache License, Version 2.0
  (the "License"); you may not use this file except in compliance with
  the License.  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */

package org.apache.commons.pool2;

import static org.junit.jupiter.api.Assertions.assertEquals;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;

/** Abstract test case for {@link ObjectPool} implementations. */
public abstract class TestKeyedObjectPool {

    protected static class FailingKeyedPooledObjectFactory {
        private final List<MethodCall> methodCalls = new ArrayList<>();
        private int count;
        private boolean makeObjectFail;
        private boolean activateObjectFail;
        private boolean validateObjectFail;
        private boolean passivateObjectFail;
        private boolean destroyObjectFail;

        public FailingKeyedPooledObjectFactory() {}

        public void activateObject(final Object key, final PooledObject<Object> obj)
                throws Exception {
            methodCalls.add(MethodCall.MethodCall0("activateObject", key, obj.getObject()));
            if (activateObjectFail) {
                throw new PrivateException("activateObject");
            }
        }

        public void destroyObject(final Object key, final PooledObject<Object> obj)
                throws Exception {
            methodCalls.add(MethodCall.MethodCall0("destroyObject", key, obj.getObject()));
            if (destroyObjectFail) {
                throw new PrivateException("destroyObject");
            }
        }

        public int getCurrentCount() {
            return count;
        }

        public List<MethodCall> getMethodCalls() {
            return methodCalls;
        }

        public boolean isActivateObjectFail() {
            return activateObjectFail;
        }

        public boolean isDestroyObjectFail() {
            return destroyObjectFail;
        }

        public boolean isMakeObjectFail() {
            return makeObjectFail;
        }

        public boolean isPassivateObjectFail() {
            return passivateObjectFail;
        }

        public boolean isValidateObjectFail() {
            return validateObjectFail;
        }

        public void passivateObject(final Object key, final PooledObject<Object> obj)
                throws Exception {
            methodCalls.add(MethodCall.MethodCall0("passivateObject", key, obj.getObject()));
            if (passivateObjectFail) {
                throw new PrivateException("passivateObject");
            }
        }

        public void reset() {
            count = 0;
            getMethodCalls().clear();
            setMakeObjectFail(false);
            setActivateObjectFail(false);
            setValidateObjectFail(false);
            setPassivateObjectFail(false);
            setDestroyObjectFail(false);
        }

        public void setActivateObjectFail(final boolean activateObjectFail) {
            this.activateObjectFail = activateObjectFail;
        }

        public void setCurrentCount(final int count) {
            this.count = count;
        }

        public void setDestroyObjectFail(final boolean destroyObjectFail) {
            this.destroyObjectFail = destroyObjectFail;
        }

        public void setMakeObjectFail(final boolean makeObjectFail) {
            this.makeObjectFail = makeObjectFail;
        }

        public void setPassivateObjectFail(final boolean passivateObjectFail) {
            this.passivateObjectFail = passivateObjectFail;
        }

        public void setValidateObjectFail(final boolean validateObjectFail) {
            this.validateObjectFail = validateObjectFail;
        }

        public boolean validateObject(final Object key, final PooledObject<Object> obj) {
            final MethodCall call = MethodCall.MethodCall0("validateObject", key, obj.getObject());
            methodCalls.add(call);
            if (validateObjectFail) {
                throw new PrivateException("validateObject");
            }
            final boolean r = true;
            call.returned(Boolean.valueOf(r));
            return r;
        }
    }

    protected static final String KEY = "key";

    private KeyedObjectPool<Object, Object> pool;

    private final Integer ZERO = Integer.valueOf(0);

    private final Integer ONE = Integer.valueOf(1);

    private void clear(
            final FailingKeyedPooledObjectFactory factory, final List<MethodCall> expectedMethods) {
        factory.getMethodCalls().clear();
        expectedMethods.clear();
    }

    /**
     * Return what we expect to be the n<sup>th</sup> object (zero indexed) created by the pool for
     * the given key.
     *
     * @param key Key for the object to be obtained
     * @param n index of the object to be obtained
     * @return the requested object
     */
    protected abstract Object getNthObject(Object key, int n);

    protected abstract boolean isFifo();

    protected abstract boolean isLifo();

    /**
     * Creates an {@link KeyedObjectPool} instance that can contain at least <i>minCapacity</i> idle
     * and active objects, or throw {@link IllegalArgumentException} if such a pool cannot be
     * created.
     *
     * @param minCapacity Minimum capacity of the pool to create
     * @return the newly created keyed object pool
     */
    protected abstract KeyedObjectPool<Object, Object> makeEmptyPool0(int minCapacity);

    /**
     * Creates an {@code KeyedObjectPool} with the specified factory. The pool should be in a
     * default configuration and conform to the expected behaviors described in {@link
     * KeyedObjectPool}. Generally speaking there should be no limits on the various object counts.
     *
     * @param factory Factory to use to associate with the pool
     * @return The newly created empty pool
     */
    protected abstract KeyedObjectPool<Object, Object> makeEmptyPool1(
            KeyedPooledObjectFactory<Object, Object> factory);

    protected abstract Object makeKey(int n);

    private void reset(
            final KeyedObjectPool<Object, Object> pool,
            final FailingKeyedPooledObjectFactory factory,
            final List<MethodCall> expectedMethods)
            throws Exception {
        pool.clear0();
        clear(factory, expectedMethods);
        factory.reset();
    }

    @AfterEach
    public void tearDown() {
        pool = null;
    }

    @Test
    public void testBaseAddObject() throws Exception {
        try {
            pool = makeEmptyPool0(3);
        } catch (final UnsupportedOperationException uoe) {
            return; // skip this test if unsupported
        }
        final Object key = makeKey(0);
        try {
            assertEquals(0, pool.getNumIdle0());
            assertEquals(0, pool.getNumActive0());
            assertEquals(0, pool.getNumIdle1(key));
            assertEquals(0, pool.getNumActive1(key));
            pool.addObject(key);
            assertEquals(1, pool.getNumIdle0());
            assertEquals(0, pool.getNumActive0());
            assertEquals(1, pool.getNumIdle1(key));
            assertEquals(0, pool.getNumActive1(key));
            final Object obj = pool.borrowObject(key);
            assertEquals(getNthObject(key, 0), obj);
            assertEquals(0, pool.getNumIdle0());
            assertEquals(1, pool.getNumActive0());
            assertEquals(0, pool.getNumIdle1(key));
            assertEquals(1, pool.getNumActive1(key));
            pool.returnObject(key, obj);
            assertEquals(1, pool.getNumIdle0());
            assertEquals(0, pool.getNumActive0());
            assertEquals(1, pool.getNumIdle1(key));
            assertEquals(0, pool.getNumActive1(key));
        } catch (final UnsupportedOperationException e) {
            return; // skip this test if one of those calls is unsupported
        } finally {
            pool.close();
        }
    }

    @Test
    public void testBaseBorrow() throws Exception {
        try {
            pool = makeEmptyPool0(3);
        } catch (final UnsupportedOperationException uoe) {
            return; // skip this test if unsupported
        }
        final Object keya = makeKey(0);
        final Object keyb = makeKey(1);
        assertEquals(getNthObject(keya, 0), pool.borrowObject(keya), "1");
        assertEquals(getNthObject(keyb, 0), pool.borrowObject(keyb), "2");
        assertEquals(getNthObject(keyb, 1), pool.borrowObject(keyb), "3");
        assertEquals(getNthObject(keya, 1), pool.borrowObject(keya), "4");
        assertEquals(getNthObject(keyb, 2), pool.borrowObject(keyb), "5");
        assertEquals(getNthObject(keya, 2), pool.borrowObject(keya), "6");
        pool.close();
    }

    @Test
    public void testBaseBorrowReturn() throws Exception {
        try {
            pool = makeEmptyPool0(3);
        } catch (final UnsupportedOperationException uoe) {
            return; // skip this test if unsupported
        }
        final Object keya = makeKey(0);
        Object obj0 = pool.borrowObject(keya);
        assertEquals(getNthObject(keya, 0), obj0);
        Object obj1 = pool.borrowObject(keya);
        assertEquals(getNthObject(keya, 1), obj1);
        Object obj2 = pool.borrowObject(keya);
        assertEquals(getNthObject(keya, 2), obj2);
        pool.returnObject(keya, obj2);
        obj2 = pool.borrowObject(keya);
        assertEquals(getNthObject(keya, 2), obj2);
        pool.returnObject(keya, obj1);
        obj1 = pool.borrowObject(keya);
        assertEquals(getNthObject(keya, 1), obj1);
        pool.returnObject(keya, obj0);
        pool.returnObject(keya, obj2);
        obj2 = pool.borrowObject(keya);
        if (isLifo()) {
            assertEquals(getNthObject(keya, 2), obj2);
        }
        if (isFifo()) {
            assertEquals(getNthObject(keya, 0), obj2);
        }
        obj0 = pool.borrowObject(keya);
        if (isLifo()) {
            assertEquals(getNthObject(keya, 0), obj0);
        }
        if (isFifo()) {
            assertEquals(getNthObject(keya, 2), obj0);
        }
        pool.close();
    }

    @Test
    public void testBaseClear() throws Exception {
        try {
            pool = makeEmptyPool0(3);
        } catch (final UnsupportedOperationException uoe) {
            return; // skip this test if unsupported
        }
        final Object keya = makeKey(0);
        assertEquals(0, pool.getNumActive1(keya));
        assertEquals(0, pool.getNumIdle1(keya));
        final Object obj0 = pool.borrowObject(keya);
        final Object obj1 = pool.borrowObject(keya);
        assertEquals(2, pool.getNumActive1(keya));
        assertEquals(0, pool.getNumIdle1(keya));
        pool.returnObject(keya, obj1);
        pool.returnObject(keya, obj0);
        assertEquals(0, pool.getNumActive1(keya));
        assertEquals(2, pool.getNumIdle1(keya));
        pool.clear1(keya);
        assertEquals(0, pool.getNumActive1(keya));
        assertEquals(0, pool.getNumIdle1(keya));
        final Object obj2 = pool.borrowObject(keya);
        assertEquals(getNthObject(keya, 2), obj2);
        pool.close();
    }

    @Test
    public void testBaseInvalidateObject() throws Exception {
        try {
            pool = makeEmptyPool0(3);
        } catch (final UnsupportedOperationException uoe) {
            return; // skip this test if unsupported
        }
        final Object keya = makeKey(0);
        assertEquals(0, pool.getNumActive1(keya));
        assertEquals(0, pool.getNumIdle1(keya));
        final Object obj0 = pool.borrowObject(keya);
        final Object obj1 = pool.borrowObject(keya);
        assertEquals(2, pool.getNumActive1(keya));
        assertEquals(0, pool.getNumIdle1(keya));
        pool.invalidateObject0(keya, obj0);
        assertEquals(1, pool.getNumActive1(keya));
        assertEquals(0, pool.getNumIdle1(keya));
        pool.invalidateObject0(keya, obj1);
        assertEquals(0, pool.getNumActive1(keya));
        assertEquals(0, pool.getNumIdle1(keya));
        pool.close();
    }

    @Test
    public void testBaseNumActiveNumIdle() throws Exception {
        try {
            pool = makeEmptyPool0(3);
        } catch (final UnsupportedOperationException uoe) {
            return; // skip this test if unsupported
        }
        final Object keya = makeKey(0);
        assertEquals(0, pool.getNumActive1(keya));
        assertEquals(0, pool.getNumIdle1(keya));
        final Object obj0 = pool.borrowObject(keya);
        assertEquals(1, pool.getNumActive1(keya));
        assertEquals(0, pool.getNumIdle1(keya));
        final Object obj1 = pool.borrowObject(keya);
        assertEquals(2, pool.getNumActive1(keya));
        assertEquals(0, pool.getNumIdle1(keya));
        pool.returnObject(keya, obj1);
        assertEquals(1, pool.getNumActive1(keya));
        assertEquals(1, pool.getNumIdle1(keya));
        pool.returnObject(keya, obj0);
        assertEquals(0, pool.getNumActive1(keya));
        assertEquals(2, pool.getNumIdle1(keya));

        assertEquals(0, pool.getNumActive1("xyzzy12345"));
        assertEquals(0, pool.getNumIdle1("xyzzy12345"));

        pool.close();
    }

    @Test
    public void testBaseNumActiveNumIdle2() throws Exception {
        try {
            pool = makeEmptyPool0(6);
        } catch (final UnsupportedOperationException uoe) {
            return; // skip this test if unsupported
        }
        final Object keya = makeKey(0);
        final Object keyb = makeKey(1);
        assertEquals(0, pool.getNumActive0());
        assertEquals(0, pool.getNumIdle0());
        assertEquals(0, pool.getNumActive1(keya));
        assertEquals(0, pool.getNumIdle1(keya));
        assertEquals(0, pool.getNumActive1(keyb));
        assertEquals(0, pool.getNumIdle1(keyb));

        final Object objA0 = pool.borrowObject(keya);
        final Object objB0 = pool.borrowObject(keyb);

        assertEquals(2, pool.getNumActive0());
        assertEquals(0, pool.getNumIdle0());
        assertEquals(1, pool.getNumActive1(keya));
        assertEquals(0, pool.getNumIdle1(keya));
        assertEquals(1, pool.getNumActive1(keyb));
        assertEquals(0, pool.getNumIdle1(keyb));

        final Object objA1 = pool.borrowObject(keya);
        final Object objB1 = pool.borrowObject(keyb);

        assertEquals(4, pool.getNumActive0());
        assertEquals(0, pool.getNumIdle0());
        assertEquals(2, pool.getNumActive1(keya));
        assertEquals(0, pool.getNumIdle1(keya));
        assertEquals(2, pool.getNumActive1(keyb));
        assertEquals(0, pool.getNumIdle1(keyb));

        pool.returnObject(keya, objA0);
        pool.returnObject(keyb, objB0);

        assertEquals(2, pool.getNumActive0());
        assertEquals(2, pool.getNumIdle0());
        assertEquals(1, pool.getNumActive1(keya));
        assertEquals(1, pool.getNumIdle1(keya));
        assertEquals(1, pool.getNumActive1(keyb));
        assertEquals(1, pool.getNumIdle1(keyb));

        pool.returnObject(keya, objA1);
        pool.returnObject(keyb, objB1);

        assertEquals(0, pool.getNumActive0());
        assertEquals(4, pool.getNumIdle0());
        assertEquals(0, pool.getNumActive1(keya));
        assertEquals(2, pool.getNumIdle1(keya));
        assertEquals(0, pool.getNumActive1(keyb));
        assertEquals(2, pool.getNumIdle1(keyb));

        pool.close();
    }
}
